accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,376 @@
++       public static TypedReference Make<T>(ref T? value);
+        // Helper used for boxed or loosely-typed cases
++       public static TypedReference Make(ref object value, Type type);
+
+        // Equivalent of __refvalue
++       public ref T GetValue<T>();
+        // Used for byref-like types or loosely-typed cases with >4 parameters
++       public readonly unsafe ref byte TargetRef { get; }
+    }
+}
+```
+
+## MethodInfo
+(these may be extension methods instead)
+
+```diff
+namespace System.Reflection
+{
+    public abstract class MethodBase
+    {
+        // Helpers for <= 4 parameters
+
+        // Note that 'default(TypedReference)' can be specified for:
+        //  - any "not used" arguments
+        //  - any trailing unused arguments (e.g. 'arg4' can be 'default' if the method only takes 3 args)
+        //  - static methods ('target' is 'default')
+        //  - 'void'-returning methods ('result' is 'default')
+
+        // We may also want to consider using 'ref byte' instead of or in addition to 'TypedReference'
+        // along with new helper methods like 'public static ref byte GetRawData(object o)'
++       [System.CLSCompliantAttribute(false)]
++       public virtual void InvokeDirect(
++           TypedReference target,
++           TypedReference result);
+
++       [System.CLSCompliantAttribute(false)]
++       public virtual void InvokeDirect(
++           TypedReference target,
++           TypedReference arg1,
++           TypedReference result);
+
++       [System.CLSCompliantAttribute(false)]
++       public virtual void InvokeDirect(
++           TypedReference target,
++           TypedReference arg1,
++           TypedReference arg2,
++           TypedReference result);
+
++       [System.CLSCompliantAttribute(false)]
++       public virtual void InvokeDirect(
++           TypedReference target,
++           TypedReference arg1,
++           TypedReference arg2,
++           TypedReference arg3,
++           TypedReference result);
+
++       [System.CLSCompliantAttribute(false)]
++       public virtual void InvokeDirect(
++           TypedReference target,
++           TypedReference arg1,
++           TypedReference arg2,
++           TypedReference arg3,
++           TypedReference arg4,
++           TypedReference result);
+
+        // Unsafe (todo: more on this; samples)
++       public virtual unsafe void InvokeDirect(
++           TypedReference target,
++           TypedReference* parameters,
++           TypedReference result);
+
+       // Also consider a safe variable-length callback that does tracking internally
+       // since we don't support a safe variable-length collection mechanism
+       // or the ability to use a Span<TypedReference>
+       // (todo: more on this?; prototype exists)
+    }
+}
+```
+
+## PropertyInfo \ FieldInfo
+```diff
+namespace System.Reflection
+{
+    public abstract class PropertyInfo
```

Comment:
We should make sure these APIs allow ASP.NET to delete this:

https://github.com/dotnet/aspnetcore/blob/4535ea1263e9a24ca8d37b7266797fe1563b8b12/src/Shared/PropertyHelper/PropertyHelper.cs